daemon/pool.go

```diff
@@ -73,123 +100,445 @@ func NewDriverPool(factory FactoryFunction) *DriverPool {
+		select {
+		case <-req.cancel:
+			return
+		case err := <-dp.spawnErr:
+			req.err <- err
+			return
+		case d := <-dp.put:
+			if dn == 0 {
+				// exactly the right amount of instances
+				returnDriver(d)
+				return
+			}
+			// bad luck - we are scaling down
+			// TODO(dennwc): add some metric to track if there are cases when the
+			//               scaling policy asks us to drain and then asks to scale
+			//               back up - we could have returned this driver to the
+			//               client instead
+			dp.killDriver(d)
+		case <-dp.rescale:
+			// worth to restart the loop
+		}
 	}
+}
 
-	dp.drivers.Lock()
-	delete(dp.drivers.byID, d.ID())
-	dp.drivers.Unlock()
-	return nil
+// drain waits until all instances are stopped. Should only be called from the
+// manager goroutine. It assumes the stop channel already triggered.
+func (dp *DriverPool) drain() {
+	defer dp.target.Set(0)
+	for {
+		d, ok := dp.peekIdle()
+		if !ok {
+			break
+		}
+		dp.killDriver(d)
+	}
+	for dp.running.Value() > 0 {
+		d := <-dp.put
+		dp.killDriver(d)
+	}
 }
 
-func (dp *DriverPool) scaling() {
-	ticker := time.NewTicker(time.Millisecond * 100)
-	defer ticker.Stop()
+// manageDrivers is the main goroutine responsible for managing drivers.
+// It will accept all client requests for drivers if there are no drivers in the idle state.
+// It will also take care of draining instances when the pool closes.
+func (dp *DriverPool) manageDrivers() {
+	defer close(dp.stopped)
+	defer dp.wg.Done()
+	defer dp.drain()
+
 	for {
 		select {
-		case <-dp.close:
-			close(dp.close)
+		case d := <-dp.put:
+			dp.setIdle(d)
+		case req := <-dp.get:
+			dp.waitOrScale(req)
+		case <-dp.rescale:
+			dp.scale()
+		case <-dp.stop:
 			return
-		case <-ticker.C:
-			dp.doScaling()
 		}
 	}
 }
 
-func (dp *DriverPool) doScaling() {
-	total := dp.stats.instances.Value()
-	ready := dp.queue.Size()
-	load := dp.stats.waiting.Value()
+// getIdle returns an idle driver from the queue. It won't check the driver status.
+// After the driver is returned, it's owned by the caller, but it still counts toward the
+// pool scaling limit. The caller should put the instance back to the pool even if the
+// driver fails.
+func (dp *DriverPool) getIdle(rctx context.Context) (Driver, error) {
+	// fast path - get an idle driver directly from the pool
+	// this function executes on the current goroutine
+	if d, ok := dp.peekIdle(); ok {
+		return d, nil
+	}
 
-	s := dp.ScalingPolicy.Scale(total, load-ready)
-	if s == total {
-		return
+	// slow path - ask the manager goroutine to pick an instance for us
+	dp.requests.Add(1)
+	defer dp.requests.Add(-1)
+
+	// ensure we can cancel our request on return
+	ctx, cancel := context.WithCancel(rctx)
+	defer cancel()
+
+	resp := make(chan Driver)
+	errc := make(chan error, 1)
+	req := driverRequest{
+		out: resp, err: errc, cancel: ctx.Done(),
+	}
+	select {
+	case <-req.cancel: // parent context cancelled
```

Comment:
Do you really mean the parent context of `rctx`? If so, I don't understand why we are checking `req.cancel` instead of `ctx.Done()` here: Isn't `req.cancel` a signal from the requestor rather than the context? 

I'm assuming I have missed something.

Reply:
It's exactly the same, but avoids a call to `Done()`. Added the comment.